File: Problemas/Grafos/Flujos/maxFlow-minCut/PoliceChase.java
//https://cses.fi/problemset/task/1695
import java.io.*;
import java.util.*;

public class PoliceChase {
	static int n;
	static ArrayList<Integer>[] adj;
	static int INF = Integer.MAX_VALUE;
	static int[][] capacity;

	@SuppressWarnings("unchecked")
	public static void main(String[] args) throws Exception {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(in.readLine());
		PrintWriter out = new PrintWriter(System.out);
		n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		adj = new ArrayList[n];
		for (int i = 0; i < n; i++)
			adj[i] = new ArrayList<>();
		capacity = new int[n][n];
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(in.readLine());
			int a = Integer.parseInt(st.nextToken()) - 1;
			int b = Integer.parseInt(st.nextToken()) - 1;
			adj[a].add(b);
			adj[b].add(a);
			capacity[a][b] = capacity[b][a] = 1;
		}
		out.println(maxflow(0, n - 1));
		boolean[] visited = dfs(0);
		for (int u = 0; u < n; u++)
			for (int v : adj[u])
				if (visited[u] && !visited[v])
					out.println((u + 1) + " " + (v + 1));
		out.flush();
	}

	static boolean[] dfs(int s) {
		boolean[] visited = new boolean[n];
		Stack<Integer> Q = new Stack<>();
		visited[s] = true;
		Q.push(s);
		while (!Q.isEmpty()) {
			int u = Q.pop();
			for (int v = 0; v < n; v++)
				if (capacity[u][v] > 0 && !visited[v]) {
					visited[v] = true;
					Q.push(v);
				}
		}
		return visited;
	}

	static int maxflow(int s, int t) {
		int[] pi = new int[n];
		int maxflow = 0;
		while (bfs(s, t, pi)) {
			int flow = INF;
			for (int u = t; u != s; u = pi[u]) {
				int p = pi[u];
				flow = Math.min(flow, capacity[p][u]);
			}
			for (int u = t; u != s; u = pi[u]) {
				int p = pi[u];
				capacity[p][u] -= flow;
				capacity[u][p] += flow;
			}
			maxflow += flow;
		}
		return maxflow;
	}

	static boolean bfs(int s, int t, int[] pi) {
		boolean[] visited = new boolean[n];
		Queue<Integer> Q = new LinkedList<>();
		Q.add(s);
		pi[s] = -1;
		visited[s] = true;
		while (!Q.isEmpty()) {
			int u = Q.poll();
			for (int v : adj[u])
				if (!visited[v] && capacity[u][v] > 0) {
					pi[v] = u;
					visited[v] = true;
					if (v == t)
						return true;
					Q.add(v);
				}
		}
		return false;
	}
	/*
	 * A minimum cut can be found after performing a maximum flow computation using
	 * the Ford-Fulkerson method. One possible minimum cut is the following: the set
	 * of all the vertices that can be reached from S in the residual graph (using
	 * edges with positive residual capacity), and the set of all the other
	 * vertices. This partition can be easily found using DFS starting at S .
	 */

}

File: Problemas/Grafos/Flujos/maxFlow-minCut/PoliceChase.cpp
//https://cses.fi/problemset/task/1695
#include <bits/stdc++.h>
using namespace std;
 
#define ll long long
#define ld long double
#define ull unsigned long long
#define pq priority_queue<int, vector<int>, greater<int>>
#define VI vector<int>
#define VVI vector<VI>
#define RAD(deg) ((deg) * PI / 180.0)
#define ones(x) __builtin_popcount(x)
#define DBG(x) cerr << #x << " = " << (x) << endl
#define For(i, a, b) for(int i = (a); i < (b); ++i)
#define Rfor(i, a, b) for(int i = (a); i >= (b); --i)
#define all(v) (v).begin(), (v).end()
#define ln '\n'
 
using pii = pair<int, int>;
using pll = pair<ll, ll>;
using pli = pair<ll, int>;
using pcc = pair<char, char>;
 
const int INF = 1e9;
const ll LINF = 1e18;
const int MOD = 1e9 + 7;
const ld PI = 2 * acos(0.0);
 
const int dx[] = {1, -1, 0, 0};
const int dy[] = {0, 0, 1, -1};
 
const int MAXN = 500;
int n;
int capacity[MAXN][MAXN];
vector<int> adj[MAXN];
 
int bfs(int s, int t, VI &parent) {
    fill(all(parent), -1);
    parent[s] = -2;
    queue<pii> q;
    q.push({s, INF});
 
    while (!q.empty()) {
        int cur = q.front().first;
        int flow = q.front().second;
        q.pop();
 
        for (int next: adj[cur]) {
            if (parent[next] == -1 && capacity[cur][next]) {
                parent[next] = cur;
                int new_flow = min(flow, capacity[cur][next]);
                if (next == t)
                    return new_flow;
                q.push({next, new_flow});
            }
        }
    }
    return 0;
}
 
int maxflow(int s, int t) {
    int flow = 0;
    VI parent(n);
    int new_flow;
 
    while (new_flow = bfs(s, t, parent)) {
        flow += new_flow;
        int cur = t;
        while (cur != s) {
            int prev = parent[cur];
            capacity[prev][cur] -= new_flow;
            capacity[cur][prev] += new_flow;
            cur = prev;
        }
    }
    return flow;
}
 
void dfs(int u, bool vis[]) {
    vis[u] = true;
    for (int v: adj[u]) {
        if (!vis[v] && capacity[u][v] > 0)
            dfs(v, vis);
    }
}
 
void solve() {
    int m;
    cin >> n >> m;
 
    memset(capacity, 0, sizeof(capacity));
    For(i, 0, n)
        adj[i].clear();
 
    For(i, 0, m) {
        int a, b;
        cin >> a >> b;
        a--, b--;
        adj[a].push_back(b);
        adj[b].push_back(a);
        capacity[a][b] = 1;
        capacity[b][a] = 1;
    }
 
    int s = 0, t = n - 1;
    int min_cut = maxflow(s, t);
    cout << min_cut << ln;
    bool vis[MAXN] = {false};
    dfs(s, vis);
    For(u, 0, n) {
        for (int v: adj[u]) {
            if (vis[u] && !vis[v])
                cout << u + 1 << " " << v + 1 << ln;
        }
    }
    cout << ln;
}
 
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    int t = 1;
    //cin >> t;
    while (t--) {
        solve();
    }
    return 0;
}
	/*
	 * A minimum cut can be found after performing a maximum flow computation using
	 * the Ford-Fulkerson method. One possible minimum cut is the following: the set
	 * of all the vertices that can be reached from S in the residual graph (using
	 * edges with positive residual capacity), and the set of all the other
	 * vertices. This partition can be easily found using DFS starting at S .
	 */
